org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/editor/quickfix/WorkbenchMarkerResolutionAdapter.xtend

```diff
@@ -67,78 +66,38 @@ class WorkbenchMarkerResolutionAdapter extends WorkbenchMarkerResolution {
 	}
 
 	override run(IMarker[] markers, IProgressMonitor monitor) {
-		new WorkspaceModifyOperation() {
-			override protected execute(
-				IProgressMonitor monitor) throws CoreException, InvocationTargetException, InterruptedException {
-				monitor.beginTask("Applying resolutions", markers.size)
-				val grouped = markers.groupBy[resource.project]
-				collectResolutions(monitor, grouped).forEach [ proj, resolutions |
-					val serializer = serializerProvider.get()
-					val converter = converterFactory.create(primaryResolution.label, null, issueAcceptor)
-					monitor.taskName = "Applying resolution"
-					resolutions.forEach [
-						run(it.key, it.value, serializer, monitor)
-					]
-					monitor.internalWorked(1)
-					serializer.applyModifications(converter)
-					val ltkChange = converter.change
-					ltkChange.initializeValidationData(monitor)
-					new PerformChangeOperation(ltkChange).run(monitor)
-				]
-				monitor.done
-			}
-		}.run(monitor)
-	}
-
-	def collectResolutions(IProgressMonitor monitor, Map<IProject, List<IMarker>> markersByProject) {
-		val result = newLinkedHashMap
-		markersByProject.forEach [ proj, markers |
-			val resSet = resSetProvider.get(proj)
-			result.put(proj, markers.map [ marker |
-				if (monitor.isCanceled) {
-					throw new OperationCanceledException()
-				}
-				return marker.resolution(resSet)
-			].filterNull.toList)
-		]
-		return result
+		monitor.beginTask("Applying resolutions", markers.size)
+		for (g : markers.groupBy[resource.project].entrySet) {
+			val batch = batchModificationProvider.get
+			batch.project = g.key
+			val markersInProject = g.value
+			monitor.cancelIfNeeded
```

Comment:
Maybe this is overly eager cancellation checking in this block, e.g. after invocation of a simple getter or some lazy iterable creation. 
On the other hand, we should probably report real progress, otherwise the ui would appear to be frozen.

Reply:
Yes, reporting worked progress would be good